performance/gcd_bench.cpp

```diff
@@ -0,0 +1,215 @@
+std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& get_test_vector(unsigned bits)
+{
+   static std::map<unsigned, std::tuple<std::vector<T>, std::vector<T>, std::vector<T> > > data;
+
+   std::tuple<std::vector<T>, std::vector<T>, std::vector<T> >& result = data[bits];
+
+   if (std::get<0>(result).size() == 0)
+   {
+      mt19937                     mt;
+      uniform_int_distribution<T> ui(T(1) << (bits - 1), T(1) << bits);
+
+      std::vector<T>& a = std::get<0>(result);
+      std::vector<T>& b = std::get<1>(result);
+      std::vector<T>& c = std::get<2>(result);
+
+      for (unsigned i = 0; i < 1000; ++i)
+      {
+         a.push_back(ui(mt));
+         b.push_back(ui(mt));
+         if (b.back() > a.back())
+            b.back().swap(a.back());
+         c.push_back(0);
+      }
+   }
+   return result;
+}
+
+template <class T>
+std::vector<T>& get_test_vector_a(unsigned bits)
+{
+   return std::get<0>(get_test_vector<T>(bits));
+}
+template <class T>
+std::vector<T>& get_test_vector_b(unsigned bits)
+{
+   return std::get<1>(get_test_vector<T>(bits));
+}
+template <class T>
+std::vector<T>& get_test_vector_c(unsigned bits)
+{
+   return std::get<2>(get_test_vector<T>(bits));
+}
+
+
+template <typename T>
+static void BM_gcd_old(benchmark::State& state)
+{
+   int                         bits = state.range(0);
+
+   std::vector<T>& a = get_test_vector_a<T>(bits);
+   std::vector<T>& b = get_test_vector_b<T>(bits);
+   std::vector<T>& c = get_test_vector_c<T>(bits);
+
+   for (auto _ : state)
+   {
+      for (unsigned i = 0; i < a.size(); ++i)
+         eval_gcd_old(c[i].backend(), a[i].backend(), b[i].backend());
+   }
+   state.SetComplexityN(bits);
+}
+
+template <typename T>
+static void BM_gcd_current(benchmark::State& state)
+{
+   int                         bits = state.range(0);
+
+   std::vector<T>& a = get_test_vector_a<T>(bits);
+   std::vector<T>& b = get_test_vector_b<T>(bits);
+   std::vector<T>& c = get_test_vector_c<T>(bits);
+
+   for (auto _ : state)
+   {
+      for (unsigned i = 0; i < a.size(); ++i)
+         eval_gcd(c[i].backend(), a[i].backend(), b[i].backend());
+   }
+   state.SetComplexityN(bits);
+}
+
+constexpr unsigned lower_range = 512;
+constexpr unsigned upper_range = 1 << 15;
+
+BENCHMARK_TEMPLATE(BM_gcd_old, cpp_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_current, cpp_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_old, cpp_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_current, mpz_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_current, mpz_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
```

Comment:
I think now that this is finished, we can remove the duplicated performance tests since you'd demonstrated an order of magnitude speedup. It's a pretty unambiguous win now!

Reply:
Nod.